misc/multistream-select/src/protocol/mod.rs

```diff
@@ -20,47 +20,125 @@
 
 //! Contains lower-level structs to handle the multistream protocol.
 
+const MSG_MULTISTREAM_1_0: &[u8] = b"/multistream/1.0.0\n";
+const MSG_PROTOCOL_NA: &[u8] = b"na\n";
+const MSG_LS: &[u8] = b"ls\n";
+
 mod dialer;
 mod error;
 mod listener;
 
-const MULTISTREAM_PROTOCOL_WITH_LF: &[u8] = b"/multistream/1.0.0\n";
-
 pub use self::dialer::{Dialer, DialerFuture};
 pub use self::error::MultistreamSelectError;
 pub use self::listener::{Listener, ListenerFuture};
 
+use bytes::{BytesMut, BufMut};
+use unsigned_varint as uvi;
+
+pub enum Header {
+    Multistream10
+}
+
+impl Header {
+    fn encode(&self, dest: &mut BytesMut) {
+        match self {
+            Header::Multistream10 => {
+                dest.reserve(MSG_MULTISTREAM_1_0.len());
+                dest.put(MSG_MULTISTREAM_1_0);
+            }
+        }
+    }
+}
+
 /// Message sent from the dialer to the listener.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub enum DialerToListenerMessage<N> {
+pub enum Request<N> {
     /// The dialer wants us to use a protocol.
     ///
     /// If this is accepted (by receiving back a `ProtocolAck`), then we immediately start
     /// communicating in the new protocol.
-    ProtocolRequest {
+    Protocol {
         /// Name of the protocol.
         name: N
     },
 
     /// The dialer requested the list of protocols that the listener supports.
-    ProtocolsListRequest,
+    ListProtocols,
 }
 
+impl<N: AsRef<[u8]>> Request<N> {
+    fn encode(&self, dest: &mut BytesMut) -> Result<(), MultistreamSelectError> {
+        match self {
+            Request::Protocol { name } => {
+                if !name.as_ref().starts_with(b"/") {
```

Comment:
Well, strictly speaking no, it is just a convention, the ["spec"](https://github.com/libp2p/specs/blob/master/connections/README.md#protocol-negotiation) says

> [..] While any string can be used, the conventional format is a path-like structure containing a short name and a version number, separated by / characters. [..]

but it doesn't even bother saying what a "string" here is in terms of encoding, so it sounds like a sentence quickly extrapolated from a go or js implementation. Practically,  the requirement for the leading `/` did not change in this PR - this "restriction" was [already](https://github.com/libp2p/rust-libp2p/blob/master/misc/multistream-select/src/protocol/dialer.rs#L182) [there](https://github.com/libp2p/rust-libp2p/blob/master/misc/multistream-select/src/protocol/listener.rs#L203) and is needed [for](https://github.com/libp2p/rust-libp2p/blob/master/misc/multistream-select/src/protocol/dialer.rs#L115) [decoding](https://github.com/libp2p/rust-libp2p/blob/master/misc/multistream-select/src/protocol/listener.rs#L108). Personally, I'm very much in favor of stricter parsing of protocol names reflecting the "conventional format", based on a simple grammar (e.g. a protocol name must start with a `/`, followed by a sequence of alphanumeric strings including dots and hyphens, separated by `/`).

Reply:
> > [..] While any string can be used, the conventional format is a path-like structure containing a short name and a version number, separated by / characters. [..]
> 
> but it doesn't even bother saying what a "string" here is in terms of encoding

I stand corrected, it does state further down that

> Messages are sent encoded as UTF-8 byte strings, [..]

which implies utf-8 encoding for protocol names. This is unfortunately also not enforced by this implementation since the interface rests on `AsRef<[u8]>`.